client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -17,123 +17,113 @@
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
 use libp2p::core::multiaddr::{Multiaddr, Protocol};
-use std::collections::HashMap;
 
+use lru::LruCache;
 use sc_network::PeerId;
 use sp_authority_discovery::AuthorityId;
+use std::collections::{hash_map::Entry, HashMap, HashSet};
 
 /// Cache for [`AuthorityId`] -> [`Vec<Multiaddr>`] and [`PeerId`] -> [`AuthorityId`] mappings.
 pub(super) struct AddrCache {
-	// The addresses found in `authority_id_to_addresses` are guaranteed to always match
-	// the peerids found in `peer_id_to_authority_id`. In other words, these two hashmaps
-	// are similar to a bi-directional map.
-	authority_id_to_addresses: HashMap<AuthorityId, Vec<Multiaddr>>,
-	peer_id_to_authority_id: HashMap<PeerId, AuthorityId>,
+	/// The addresses found in `authority_id_to_addresses` are guaranteed to always match
+	/// the peerids found in `peer_id_to_authority_id`. In other words, these two maps
+	/// are similar to a bi-directional map. As both are lru maps, we ensure that only inserts
+	/// update the position in the lru and any other access doesn't update the position of a
+	/// requested item.
+	authority_id_to_addresses: LruCache<AuthorityId, HashSet<Multiaddr>>,
+	peer_id_to_authority_id: HashMap<PeerId, HashSet<AuthorityId>>,
+	/// The maximum number of authority ids that should be cached.
+	max_authority_ids: usize,
 }
 
 impl AddrCache {
 	pub fn new() -> Self {
 		AddrCache {
-			authority_id_to_addresses: HashMap::new(),
-			peer_id_to_authority_id: HashMap::new(),
+			authority_id_to_addresses: LruCache::unbounded(),
+			peer_id_to_authority_id: HashMap::default(),
+			max_authority_ids: 0,
```

Comment:
if we start with 0, won't we lose entries initially before calling `refill_pending_lookups_queue`?